lib/floe/workflow.rb

```diff
@@ -119,8 +119,14 @@ def run_nonblock
     def step_nonblock
       return Errno::EPERM if end?
 
-      step_next
-      current_state.run_nonblock!.tap { end_workflow }
+      result = current_state.run_nonblock!
+
+      if context.state_finished?
+        step_next
+        end_workflow
```

Comment:
I wanted to drop the state_finished? from end_workflow but it is called in a bunch of specs